File: Combinatorics_&_Permutations/Q1. Permutation Sequence/Q1_PermutationSequence.java
import java.util.*;

class Solution {
    public String getPermutation(int n, int k) {
        List<Integer> numbers = new ArrayList<>();  // List to hold numbers 1 to n
        int[] fact = new int[n + 1];                // Array to store factorials

        fact[0] = 1;  // 0! = 1
        // Compute all factorials up to n!
        for (int i = 1; i <= n; i++) {
            fact[i] = fact[i - 1] * i;
            numbers.add(i);  // Add numbers 1, 2, 3, ..., n
        }

        k--;  // Convert k to 0-based index
        StringBuilder sb = new StringBuilder();  // To build the permutation string

        // Build permutation by selecting one digit at a time
        for (int i = n; i >= 1; i--) {
            int index = k / fact[i - 1];  // Determine which number to pick
            sb.append(numbers.get(index));  // Add selected number to result
            numbers.remove(index);          // Remove used number
            k %= fact[i - 1];               // Update k for remaining digits
        }

        return sb.toString();  // Return the kth permutation as string
    }
}

File: Combinatorics_&_Permutations/Q1. Permutation Sequence/Q1_PermutationSequence.py
class Solution(object):
    def getPermutation(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """
        import math  # To use factorial function
        numbers = [str(i) for i in range(1, n + 1)]  # List of numbers as strings
        k -= 1  # Convert k to 0-based index
        result = []  # To store the permutation result

        # Build the kth permutation
        for i in range(n, 0, -1):
            fact = math.factorial(i - 1)  # Compute (i-1)!
            index = k // fact  # Find which number to choose
            result.append(numbers[index])  # Add that number to result
            numbers.pop(index)  # Remove used number
            k %= fact  # Update k for next position

        return ''.join(result)  # Join list into string and return
